migration_scripts/bulk_load/src/footprint.clj

```diff
@@ -0,0 +1,263 @@
+
+  In that case, we will use no extra storage for primary keys"
+  [tables]
+  (let [no-pkey #(dissoc % :pkey)]
+    (->> tables
+         (map (fn [[table stats]]
+                [table (->> stats (map no-pkey) (into []))]))
+         (into {}))))
+
+;; Pruning ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn pruned
+  "Simulate pruning to only contains the most recent `days` worth of data.
+
+  Pruning for epoch partitioned tables is simulated by removing those
+  tables. Pruning for other tables that get new data every day is
+  simulated by removing a proportional number of elements."
+  [days tables]
+  (let [parts (partitions tables)
+        ratio (/ days parts)
+
+        ;; Simulate pruning by dropping all but the last `days` partitions.
+        drop-parts #(->> % reverse (take days) reverse (into []))
+
+        ;; Simulate pruning by reducing the footprint of all the parts of this
+        ;; table uniformly by the ratio between the number of days to prune and
+        ;; the number of partitions in the table.
+        prorate
+        #(->> %
+              (map (fn [{:keys [self pkey idx toast tuples]}]
+                     (cond-> {}
+                       self   (assoc :self   (long (* self   ratio)))
+                       pkey   (assoc :pkey   (long (* pkey   ratio)))
+                       idx    (assoc :idx    (long (* idx    ratio)))
+                       toast  (assoc :toast  (long (* toast  ratio)))
+                       tuples (assoc :tuples (long (* tuples ratio))))))
+              (into []))]
+    reduce
+    (as-> tables $
+      ;; Simulate pruning in these tables by dropping partitions
+      (reduce #(update %1 %2 drop-parts) $
+              [:objects-history
+               :transactions
+               :events])
+
+      ;; Simulate pruning in these tables by removing a proportional number of
+      ;; rows.
+      (reduce #(update %1 %2 prorate) $
+              [:checkpoints
+               :epochs
+
+               :event-emit-module
+               :event-emit-package
+               :event-senders
+               :event-struct-instantiation
+               :event-struct-module
+               :event-struct-name
+               :event-struct-package
+
+               :objects-version
+
+               :tx-calls-fun
+               :tx-calls-mod
+               :tx-calls-pkg
+               :tx-changed-objects
+               :tx-digests
+               :tx-input-objects
+               :tx-kinds
+               :tx-recipients
+               :tx-senders]))))
+
+(defn efficient-pruning-scan
+  "Some tables require an extra index to be added to support efficiently
+  selecting the values that should be pruned. Without it, pruning must
+  scan over ranges manually.
+
+  This modifier simulates adding the extra index -- it budgets one
+  long word for the sequence number (the index key) and one long word
+  for the pointer into the heap."
+  [tables]
+  (let [idx-overhead
```

Comment:
this would be called after simulating pruning a table, to determine the overhead needed?

Reply:
It can be called at any time (before or after the function that simulates pruning) but yes, it's designed to add the cost of the extra indices that are necessary to support pruning.